src/main.rs

```diff
@@ -114,15 +117,35 @@ async fn main() -> Result<(), Box<dyn Error>> {
 
     let server = Server::bind(&address).serve(service);
 
+    let graceful = server.with_graceful_shutdown(shutdown_signal());
+
     info!("Listening on http://{}", address);
 
-    if let Err(why) = server.await {
+    if let Err(why) = graceful.await {
         error!("Fatal server error: {}", why);
     }
 
     Ok(())
 }
 
+async fn shutdown_signal() {
```

Comment:
Could we make these different functions, or format in the function better? It feels hard to read.